drivers/interrupt_controller/intc_renesas_ra_icu.c

```diff
@@ -26,99 +30,112 @@ enum {
 	IELSRn_OFFSET = 0x300,
 };
 
-int ra_icu_query_exists_irq(uint32_t event)
+static inline unsigned int event_to_irq(unsigned int event)
 {
-	for (uint32_t i = 0; i < CONFIG_NUM_IRQS; i++) {
-		uint32_t els = sys_read32(IELSRn_REG(i)) & UINT8_MAX;
-
-		if (event == els) {
-			return i;
-		}
+	if (event >= sizeof(function_irq_table)) {
+		return -1;
 	}
 
-	return -EINVAL;
+	return function_irq_table[event];
+}
+
+#ifdef CONFIG_ARM_CUSTOM_INTERRUPT_CONTROLLER
+void arm_irq_priority_set(unsigned int irq, unsigned int prio, uint32_t flags);
+
+void z_soc_irq_init(void)
+{
 }
 
-int ra_icu_query_available_irq(uint32_t event)
+void z_soc_irq_enable(unsigned int event)
 {
-	int irq = -EINVAL;
+	const unsigned int irq = event_to_irq(event);
 
```

Comment:
This is a point to discuss.

In zephyr design, the `IRQ Number` seems to have two meanings,
One is the interruption uses a which physical line.
The other is the function type of interruption.

The two are the same as usual in Zephyr.
There has never been a need to consider this before.

However, the two meanings are entirely different in Renesas RA's case.

What line corresponds to an interruption is an NVIC issue, and what function type of interruption is the ICU's.
DeviceTree is typical case of IRQ number treat as function type.
Your point is correct.
But I think this situation is literally correct but has a semantic problem.

From another perspective, the ICU acts as an abstraction layer for interruptions, hiding the physical interruption line. High-layer implementation only needs to consider the function type of interruption.

This PR intends to provide a driver that correctly handles the abstraction structure of raw IRQ.
The ICU provides an abstraction that hides the physical IRQ number; I think it is structurally correct to specify the event number with the `abstracted IRQ number,` that is, the event number, to the ICU.
There seems to be a difference between the Zephyr design and the ICU's functionality.

Reply:
@soburi , In case we move on with this solution for RA ICU, I think the document should be update in case we enable the CONFIG_SW_ISR_TABLE_USE_FUNCTION_NUMBER

Reply:
After further consideration, I decided to leave the irq operations as they were, and create an RA ICU driver that would convert between id and irq no as a wrapper for the irq operations.
Even though it's called a driver, there is no common API, so it's just a simple addition of functionality.
I don't think this will be a problem.